shared/middleware/files_api.rb

```diff
@@ -786,15 +791,18 @@ def files_put_file(encrypted_channel_id, filename, body)
     s3_prefix = "#{METADATA_PATH}/#{filename}"
     if THUMBNAIL_FILENAME == filename
       begin
-        temp_url = FileBucket.new.make_temporary_public_url(encrypted_channel_id, s3_prefix)
-        rating = ImageModeration.rate_image(temp_url)
-        if %i(adult racy).include? rating
-          # Incrementing abuse score by 15 to differentiate from manually reported projects
-          new_score = StorageApps.new(storage_id('user')).increment_abuse(encrypted_channel_id, 15)
-          FileBucket.new.replace_abuse_score(encrypted_channel_id, s3_prefix, new_score)
-          response.headers['x-cdo-content-rating'] = rating.to_s
-          cache_for 1.hour
-          not_found
+        storage_apps = StorageApps.new(storage_id('user'))
+        if MODERATE_THUMBNAILS_FOR_PROJECT_TYPES.include? project_type(encrypted_channel_id, storage_apps)
+          temp_url = FileBucket.new.make_temporary_public_url(encrypted_channel_id, s3_prefix)
+          rating = ImageModeration.rate_image(temp_url)
+          if %i(adult racy).include? rating
+            # Incrementing abuse score by 15 to differentiate from manually reported projects
+            new_score = storage_apps.increment_abuse(encrypted_channel_id, 15)
+            FileBucket.new.replace_abuse_score(encrypted_channel_id, s3_prefix, new_score)
+            response.headers['x-cdo-content-rating'] = rating.to_s
+            cache_for 1.hour
+            not_found
+          end
```

Comment:
Note: If you find the complexity of this method a little alarming, so do I.  The reason I haven't done more refactoring here is that I'm still planning to pivot to a different structure for interacting with the Image Moderation service, [as described here](https://github.com/code-dot-org/code-dot-org/pull/21812#issuecomment-380916601) - that will more or less turn this method inside out.

The good news is, all the tests I've written will still be valid (although their VCR cassettes will probably need regenerating).

Reply:
Did this follow up ever happen? Still something we should prioritize?

Reply:
This did happen - if you check out the latest `ImageModeration` code you'll see that it now sends binary image data to the moderation provider instead of just sending them a temporary public URL.